[DPE-1526] Implement a robust URI cleaning process for SYNSTAGE's `outdir` parameter

jaymedina opened the pull request:
# **Problem:**

Issue raised [here](https://sagebionetworks.slack.com/archives/C048RL0RY74/p1765903284465309) & notes from investigation are found [here](https://sagebionetworks.jira.com/browse/DPE-1526?focusedCommentId=279774).

The `outdir_clean` parameter is built on the assumption that the URI provided (either the default `outdir` or an `outdir` provided by the user) is formatted such that there is NO URI authority, and only a URI scheme and path. See the _Execution Log_ of this [workflow run](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/3hl6mtL3klDQHD) for why the original path configuration was able to work for default paths. **Long story short, when the workflow uses the default `outdir` parameter, this `outdir` is built off the `workDir` of your Seqera Platform workspace, which seems to have a leading `/`. The extra `/` after `s3://` omits a URI authority, and treats the bucket name as part of the path.**

Results from the example workflow run, which uses the default `outdir` built off `workDir`:
```
params.input: s3://example-dev-project-tower-scratch/work/my_test_dataset.csv
params.outdir: s3:///example-dev-project-tower-scratch//work/synstage/
uri.path: /example-dev-project-tower-scratch//work/synstage/
uri.authority: null
params.outdir_clean: s3://example-dev-project-tower-scratch/work/synstage
```

When users provide an `outdir` themselves, likely in the format `s3://<bucket>/<path>` this leads to the following chain of issues:
1. **On `synstage.nf`**: The `outdir_clean` is incorrectly built without using the URI authority (`<bucket>`)
2. **On `stage_file.nf`**: The `Utils.get_publish_dir` creates an incorrect `upload_location` off that `outdir_clean` parameter
3. **On `stage_file.nf`**: The AWS CLI command to copy the file from local (EC2 instance) -> to S3 is using an incorrect S3 URI, and therefore cannot place the file(s) in the appropriate location

Separately, there are a lot of instances where `/` are added manually instead of programmatically, which adds confusion to the code and makes it more error-prone.

# **Solution:**

- [x] Rebuild `outdir_clean` in a way that handles multiple different formatting cases, so that no part of the original URI is left out (scheme, authority, path).
- [x] Remove manual path concatenation where possible

# **Testing:**

For each **Result**:
1) We check that the workflow ran successfully with the correct `outdir_clean`
2) We check that the samplesheet's files were stored in the correct location on S3

---

**Case 1**: Default `outdir` = `s3:///<path>`
**Result 1**: ✅ [link](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/AyFK41012f3i8)
<img width="613" height="208" alt="image" src="https://github.com/user-attachments/assets/69e5f7f1-00c1-449b-b871-a8d2845a631f" />
<img width="872" height="407" alt="image" src="https://github.com/user-attachments/assets/9cda5cd1-0979-4bd9-9e6c-e0a69a4a7e18" />


---

**Case 2**: User-provided `outdir` = `s3://<authority>/<path>/`
**Result 2**: ✅ [link](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/g2DvnXUvtH967)
<img width="620" height="223" alt="image" src="https://github.com/user-attachments/assets/1af906c2-f50d-4f81-a8eb-afaede167163" />
<img width="1026" height="727" alt="image" src="https://github.com/user-attachments/assets/0fdf744e-2193-4947-9d0e-6c091332850c" />

---

**Case 3**: User-provided `outdir` = `s3://<authority>/<path_level1>/<path_level2>`
**Result 3**: ✅ ([link](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/5s72oBWk3tM6jh))
<img width="623" height="223" alt="image" src="https://github.com/user-attachments/assets/53added2-e8a7-4d6a-a914-9aa086afc4ac" />
<img width="786" height="445" alt="image" src="https://github.com/user-attachments/assets/526e8c10-d447-4822-b695-3d0183624055" />

---

**Case 4**: User-provided `outdir` = `s3://<authority>/<path_level1>//<path_level2>//`
**Result 4**: ✅ ([link](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/15tzSX6FPEoxt5))
<img width="617" height="223" alt="image" src="https://github.com/user-attachments/assets/73a9244d-b3d8-447f-89e2-e967c5b93621" />
<img width="755" height="430" alt="image" src="https://github.com/user-attachments/assets/d8af1547-5ba2-43dc-80b6-6dac418649a9" />

---

**Case 5**: User-provided `outdir` = `s3://<authority>/.`
**Result 5**: ✅ ([link](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/6w5FPXVX9AIiA))
<img width="635" height="219" alt="image" src="https://github.com/user-attachments/assets/f52bef19-f1d4-4a2b-9a5d-5ee14719b471" />
<img width="701" height="443" alt="image" src="https://github.com/user-attachments/assets/98ac55bb-3a4c-4ee5-bc72-8788ee5c8eac" />

---

**Case 6** (Error): User-provided `outdir` = `/<authority>/<path>`
**Result 6**: ✅ ([link](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/3AwU3bsLP5PQeo))
This should throw an error because there is no scheme in the URI

---

**Case 7** (Error): User-provided `outdir` = `s3://`
**Result 7**: ✅ ([link](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/3OAuxwDPE8B0AP))
This should throw an error because there is no authority or path in the URI

jaymedina committed:
add debug logging to figure out why outdir_clean sometimes works and sometimes doesnt

jaymedina committed:
abstract URI-cleaning to a Utils function. resolve paths instead of manually fiddling with slashes. add an extra log output for troubleshooting.

jaymedina committed:
use workDir instead of building it from scratch

jaymedina committed:
revert id_folders mapping back

jaymedina committed:
refactor clean_uri so its easy to understand the logic

jaymedina committed:
missing ) in clean_uri exception message

jaymedina committed:
remove debug log statements

jaymedina committed:
URI() already has exception handling for missing authority

jaymedina committed:
Clarify outdir description in README

rxu17 commented on lib/Utils.groovy:
Thanks for this fix! And the tests report. I am thinking we should add in unit tests even if it's not integrated in a CI/CD (it can) so that we have it in place if we ever run into this again.

I have the following simple test script I wrote just for the `clean_uri` function. I'm thinking it can go under `tests/test_utils.nf` 

You can run the tests locally by doing `nextflow run tests/test_utils.nf -lib ./lib` (once you save it to the folder I mentioned above)

I think there is groovy testing support via [spock](https://www.baeldung.com/groovy-spock) but I think the overhead for maintainence may be higher and more dependencies/plugins to add (I don't think nextflow has spock over the hood).

```groovy
/*
unit tests for Utils.groovy - no package dependencies
*/

nextflow.enable.dsl = 2

workflow {

  def failures = []

  def check = { String name, Closure test ->
    try {
      test.call()
      println "✅ ${name}"
    }
    catch (Throwable e) {
      println "❌ ${name} — ${e.class.simpleName}: ${e.message}"
      failures << [name: name, error: e]
    }
  }

  // ----------------------------
  // S3 scheme + bucket/path
  // ----------------------------
  check("clean_uri: preserves bucket/path and strips trailing slash") {
    def got = Utils.clean_uri('s3://example-bucket/some_test_dir/')
    assert got == 's3://example-bucket/some_test_dir' : "got=${got}"
  }

  check("clean_uri: removes duplicate + trailing slashes") {
    def got = Utils.clean_uri('s3://example-bucket//some_test_dir///')
    assert got == 's3://example-bucket/some_test_dir' : "got=${got}"
  }

  check("clean_uri: normalizes internal duplicate slashes") {
    def got = Utils.clean_uri('s3://example-bucket/a//b///c/')
    assert got == 's3://example-bucket/a/b/c' : "got=${got}"
  }

  // ----------------------------
  // Bucket-only forms
  // ----------------------------
  check("clean_uri: bucket-only (no trailing slash)") {
    def got = Utils.clean_uri('s3://example-bucket')
    assert got == 's3://example-bucket' : "got=${got}"
  }

  check("clean_uri: bucket-only (with trailing slash)") {
    def got = Utils.clean_uri('s3://example-bucket/')
    assert got == 's3://example-bucket' : "got=${got}"
  }

  // ----------------------------
  // file:// URI (authority is null)
  // ----------------------------
  check("clean_uri: file URI normalization (authority is null)") {
    def got = Utils.clean_uri('file:///tmp//some_dir///')
    assert got == 'file://tmp/some_dir' : "got=${got}"
  }

  check("clean_uri: s3 URI normalization (authority is null)") {
    def got = Utils.clean_uri('s3:///home/ec2-user/nf-synapse/work/synstage/')
    assert got == 's3://home/ec2-user/nf-synapse/work/synstage' : "got=${got}"
  }

  // --------------------------------------
  // Missing scheme should throw exception
  // --------------------------------------
  check("clean_uri: throws if scheme missing, outdir = /<authority>/<path>") {
    boolean threw = false
    try {
      Utils.clean_uri('/tmp/some_dir')
    } catch (Exception e) {
      threw = true
      assert e.message.contains('There must be a scheme') : "msg=${e.message}"
    }
    assert threw : "Expected exception for missing scheme"
  }

  check("clean_uri: throws if authority or path missing, outdir = s3://") {
    boolean threw = false
    try {
      Utils.clean_uri('s3://')
    } catch (Exception e) {
      threw = true
      assert e.message.contains('Expected authority at index 5') : "msg=${e.message}"
    }
    assert threw : "Expected exception for missing authority or path"
  }


  // ----------------------------
  // Summary / exit
  // ----------------------------
  if (failures) {
    println "\n========== TEST SUMMARY =========="
    println "Total: ${failures.size()} failed\n"
    failures.eachWithIndex { f, i ->
      println "${i + 1}) ${f.name}"
    }
    error "Utils tests failed (${failures.size()})."
  } else {
    println "\n✅ All Utils tests passed"
  }
}

```

Output report:

<img width="569" height="164" alt="Image" src="https://github.com/user-attachments/assets/543abc21-e1cf-4f3c-be1b-4d5aae2fb06c" />

rxu17 replied:
If the above looks okay to you (no additional tests etc), I'll add it in + add to the docs

jaymedina replied:
Good idea, yes, please feel free to add it in a branch off of this one, and we can review+merge! It would be nice to have some automated CI/CD in-place for the workflows in this repository, since it's used outside of DPE.

rxu17 reviewed (approved):
LGTM! I had a few comments

thomasyu888 reviewed (approved):
🔥 LGTM. It's not clear to me with the tests why Sashas input didn't work but maybe having this will help us understand why.

jaymedina committed:
Clarify 'outdir' default behavior in README

Updated the 'outdir' description to clarify default behavior for local and Seqera Platform runs.

jaymedina commented on README.md:
```diff
@@ -131,7 +131,7 @@ Check out the [Quickstart](#Quickstart:SYNSTAGE) section for example parameter v
 - **`entry`**: (Required) The name of the workflow to run (`synstage`). This should be the name of the workflow file in the `workflows/` directory.
 - **`input`**: (Required) A text file containing Synapse URIs (_e.g._ `syn://syn28521174`). The text file can have any format (_e.g._ a single column of Synapse URIs, a CSV/TSV sample sheet for an nf-core workflow).
-- **`outdir`**: (Optional) An output location where the Synapse files will be staged. Currently, this location must be an S3 prefix for Nextflow Tower runs. If not provided, this will default to the parent directory of the input file.
+- **`outdir`**: (Optional) An output location where the Synapse files will be staged. Currently, this location must be an S3 prefix for Nextflow Tower runs (_e.g._ `s3://bucket_name/path/to/file`). If not provided, this will default to the parent directory of the input file.
```
Done! 

https://github.com/Sage-Bionetworks-Workflows/nf-synapse/pull/15/commits/a70f1f5c482096aeba045a82df5aabb2c1645020

thomasyu888 commented:
> >  🔥 LGTM. It's not clear to me with the tests why Sashas input didn't work but maybe having this will help us understand why.
> 
> Her input wasn't working because previously the URI cleaning only concatenates the scheme (s3://) and path part of the URI (e.g: /synstage/...), but didn't consider the authority part of the URI which is typically in our case the bucket (e.g: nf-bucket). Which is why her input went from `s3://bucket-name/path-to-something` to `s3://path-to-something`

Thank you! This is helpful! I'll leave it to you to coordinate with Sasha! Thank you both!

rxu17 committed:
[DPE-1526] Add unit tests for Utils and lightweight CI/CD (#16)

* add unit tests and ci cd

* add unit testing to docs

* ignore docs changes

* remove cache

* fix tests

* add test case for s3:///path

---------

Co-authored-by: Jenny Medina <medina.stsci@gmail.com>

jaymedina commented:
@rxu17 I just merged in your PR and the tests look good + are passing ✔️ 

I'm requesting a final review before merge. Once I merge this branch into `main`, feel free to communicate this change with Sasha. Thanks!

rxu17 reviewed (approved):
LGTM!
